### PR TITLE
GPJ: first commit syn_text_pii_scanner with pipeline elements and tests

### PR DESCRIPTION
**Pull request by gpadres**
# Description

Pipeline for synthetic text PII scanner.

## How Has This Been Tested?

Tests for all functions and classes included.

### CODE REVIEW

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/pii_token_classif_models/ner_longformer_model.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            input_ids (torch.Tensor): Tensor of shape (batch_size, sequence_length).
+            attention_mask (torch.Tensor): Tensor of shape (batch_size, sequence_length).
+            labels (Optional[torch.Tensor]): Tensor of shape (batch_size, sequence_length) or None.
+
+        Returns:
+            torch.Tensor: Logits for each token, shape (batch_size, sequence_length, num_labels)
+
+        """
+        #Longformer model cannot manage more than 4096 tokens
+        input_ids = input_ids[:, 0:4096].clone()
```
Comment (johanos1) [inline leakpro/synthetic_data_attacks/syn_text_pii_scanner/pii_token_classif_models/ner_longformer_model.py:81]: Would it make sense to have a warning here in case input_ids is longer than the context window?

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/utils.py
Comment (johanos1) [inline leakpro/synthetic_data_attacks/syn_text_pii_scanner/utils.py:1]: There are lots of prints in this module. Should we replace them with the logger?

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py
Comment (johanos1) [inline leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py:1]: Nice use of pydantic. I will take inspiration from this from now on!

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            idx (int): The index of the encoded document.
+
+        Returns:
+            EncodedDoc: The encoded document at the given index.
+
+        """
+        return self.data[idx]
+
+@dataclass
+class CollatorWPadding:
```
Comment (johanos1) [inline leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py:217]: Didnt know it was possible to use dataclass in collator functions for the data loader. This makes it so much more flexible! Nice.

File: examples/synthetic_data/syn_text_pii_scanner_example.ipynb
Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: Few comments here (see ReviewNB for exact position):

- Add that datasets must be installed using pip install datasets before running the code. Note: should only be mentioned here and not added in toml as this is the only file using it.
- remove torch from the first cell.
- Please add a description of how to interpret the list of PIIs in the ori and predicted. What you want to show is that all the correct labels are found right? So the point here is that the NERLongFormer performs well.

- Similarly for the next cell, it is hard to understand what the percentiles mean. Please be a more descriptive in what the cell will show.

- In the last cell, express how to read the final printed table.

**Review by johanos1 — CHANGES_REQUESTED** [review]

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/pii_token_classif_models/ner_longformer_model.py (hunk 1 above)
Comment (gpadres) [reply leakpro/synthetic_data_attacks/syn_text_pii_scanner/pii_token_classif_models/ner_longformer_model.py:81]: Yes, i'll add it to the logger

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/utils.py
Comment (gpadres) [reply leakpro/synthetic_data_attacks/syn_text_pii_scanner/utils.py:1]: Yes

File: leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py
Comment (gpadres) [reply leakpro/synthetic_data_attacks/syn_text_pii_scanner/data_handling.py:1]: Unsure what was the inspiration, but nice to hear!

File: examples/synthetic_data/syn_text_pii_scanner_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,608 @@
+{
```
Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <p>Can we please add a description of what the example will be about? Something like:</p><p> </p><p><strong>Goal of the example</strong></p><p> </p><p><strong>Steps that will be showcased:</strong></p><ol><li>Load the TAB dataset and train a NERLongformer</li><li>Load a synthesized dataset based on X</li><li>Assess the presense of personal identifiable information (PII) that overlaps within the original and the synthetic documents</li><li>Showcase how public PIIs (PIIs present in many documents) can be excluded to focus only on the more sensitive non-public PIIs.</li></ol>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    import torch
</pre><p>remove this line, it is not used</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    from datasets import load_dataset
</pre><p>ask user to install datasets in the cell description</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <p>State that the predictions will be appended to the data object.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <p>Be more explicit in what the reader should learn. That is, the NERLongFormer, fine tuned on the TAB data, is good at identifying the entities.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <p>It is hard to understand what the distribution actually is saying. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (johanos1) [inline examples/synthetic_data/syn_text_pii_scanner_example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.    for i in sorted_sim_items[0:20]:
</pre><p>Add a line stating what the different columns are in the output.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/197/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by johanos1 — APPROVED** [review]
